File: Engine/Axis/Source/Engine/Utilities/TexCoords/AxPlanarTexCoords.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxPlanarTexCoords.h"

#include "..\AxMaths.h"

AxPlanarTexCoords::AxPlanarTexCoords(AxDeviceMesh &result, AxVector3 &planeX, AxVector3 &planeY)
{
	AxVector2 centerCorrection(0.5f, 0.5f);

	int numVertices = result.GetVertexCount();
	for (int i = 0; i < numVertices; i++)
	{
		AxVector3 vPos;
		result.GetVertexPosition(i, vPos);

		AxVector2 texCoords;
		AxMaths::GetBarycentricCoordinates(texCoords, vPos, planeX, planeY);
				
		AxVector2::Add(texCoords, texCoords, centerCorrection);
		result.SetVertexTexCoords(i, texCoords);
	}
}


AxPlanarTexCoords::~AxPlanarTexCoords(void)
{
}

File: Web/Engine/Utilities/TexCoords/AxPlanarTexCoords.js
/**
 * Creates planar mapped texture coordinates for a mesh
 * @param {AxDeviceMesh} result The mesh to map texture coordinates on to
 * @param {AxVector3} planeX The orientation and size of the mapping plane's X axis
 * @param {AxVector3} planeY The orientation and size of the mapping plane's Y axis
 * @constructor
 */
function AxPlanarTexCoords(result, planeX, planeY)
{
    var centerCorrection = new AxVector2(0.5, 0.5);

    var numVertices = result.GetVertexCount();
    for (var i = 0; i < numVertices; i++)
    {
        var vPos = new AxVector3();
        result.GetVertexPosition(i, vPos);

        var texCoords = new AxVector2();
        AxMaths.GetBarycentricCoordinates(texCoords, vPos, planeX, planeY);

        AxVector2.Add(texCoords, texCoords, centerCorrection);
        result.SetVertexTexCoords(i, texCoords);
    }
}
